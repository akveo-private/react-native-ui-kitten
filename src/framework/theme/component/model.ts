// TODO(@type): declare theme types

export type ThemeShape = any;
export type ThemedStyleShape = any;
export type StyleSheetShape = any;
