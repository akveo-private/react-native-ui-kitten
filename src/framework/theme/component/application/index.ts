export {
  ApplicationProvider,
  Props as ApplicationProviderProps,
} from './applicationProvider.component';
