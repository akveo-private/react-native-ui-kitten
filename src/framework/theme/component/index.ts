export { ThemeProvider } from './themeProvider.component';
export { withTheme } from './themeConsumer.component';
export { withThemedStyles } from './themeConsumerStyled.component';
export { ThemeType } from './type';
