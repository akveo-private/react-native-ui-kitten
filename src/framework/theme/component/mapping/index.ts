export * from './type';
export {
  MappingProvider as ThemeMappingProvider,
  Props as ThemeMappingProviderProps,
} from './mappingProvider.component';
