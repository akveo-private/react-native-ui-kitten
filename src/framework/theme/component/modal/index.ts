export * from './modal.component';
export * from './modalPanel.component';
