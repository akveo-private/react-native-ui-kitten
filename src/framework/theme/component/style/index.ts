export {
  StyleProvider,
  Props as StyleProviderProps,
} from './styleProvider.component';

export {
  styled,
  Props as StyledComponentProps,
} from './styleConsumer.component';
