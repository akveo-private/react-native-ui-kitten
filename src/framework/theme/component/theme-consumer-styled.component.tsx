import React from 'react';
import { StyleSheet } from 'react-native';
import {
  ThemeType,
  ThemedStyleType,
  StyleSheetType,
} from './type';
import {
  Consumer,
  forwardProps,
} from '../service';

export interface Props<P> extends React.ClassAttributes<P> {
  theme: ThemeType;
  themedStyle: ThemedStyleType;
}

export function withThemedStyles<P extends Props<P>>(
  Component: React.ComponentType<P>,
  createStyles: (theme: ThemeType) => StyleSheetType,
) {
  type TExcept = Exclude<keyof P, keyof Props<P>>;
  type ForwardedProps = Pick<P, TExcept>;

  class Shadow extends React.Component<ForwardedProps> {
    wrappedComponentRef = undefined;
    getWrappedInstance = undefined;

    setWrappedComponentRef = (ref) => {
      this.wrappedComponentRef = ref;
    };

    renderWrappedComponent = (theme: ThemeType) => {
      const styles = StyleSheet.create(createStyles(theme));
      return (
        <Component
          ref={this.setWrappedComponentRef}
          theme={theme}
          themedStyle={styles}
          {...this.props}
        />
      );
    };

    render() {
      return (
        <Consumer>
          {this.renderWrappedComponent}
        </Consumer>
      );
    }
  }

  const Result = Shadow;
  Result.prototype.getWrappedInstance = function getWrappedInstance() {
    const hasWrappedInstance = this.wrappedComponentRef && this.wrappedComponentRef.getWrappedInstance;
    return hasWrappedInstance ? this.wrappedComponentRef.getWrappedInstance() : this.wrappedComponentRef;
  };

  return forwardProps(Component, Result);
}
