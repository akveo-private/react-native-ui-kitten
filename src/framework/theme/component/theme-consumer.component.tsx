import React from 'react';
import {
  Consumer,
  forwardProps,
} from '../service';

export interface ThemedComponentProps<T> extends React.ClassAttributes<T> {
  theme: Object;
}

export function withTheme<T extends ThemedComponentProps<T>>(Component: React.ComponentType<T>) {
  type TExcept = Exclude<keyof T, keyof ThemedComponentProps<T>>;
  type ForwardedProps = Pick<T, TExcept>;

  class ConsumingComponent extends React.Component<ForwardedProps, {}> {
    rootComponentRef = undefined;

    getConsumingComponent = undefined;

    constructor(props) {
      super(props);
      this.setRootComponentRef = this.setRootComponentRef.bind(this);
    }

    setRootComponentRef(ref) {
      this.rootComponentRef = ref;
    }

    renderRootComponent(component: React.ComponentType<T>, theme: Object) {
      const RootComponent = component;
      return (
        <RootComponent
          ref={this.setRootComponentRef}
          theme={theme}
          {...this.props}
        />
      );
    }

    render() {
      return (
        <Consumer>
          {theme => this.renderRootComponent(Component, theme)}
        </Consumer>
      );
    }
  }

  const ResultComponent = ConsumingComponent;
  ResultComponent.prototype.getConsumingComponent = function getWrappedInstance() {
    const hasWrappedInstance = this.rootComponentRef.getConsumingComponent;
    return hasWrappedInstance ? this.rootComponentRef.getConsumingComponent() : this.rootComponentRef;
  };

  return forwardProps(ResultComponent, Component);
}
