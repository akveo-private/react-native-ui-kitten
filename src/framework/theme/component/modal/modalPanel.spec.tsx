import React from 'react';
import {
  fireEvent,
  render,
} from 'react-native-testing-library';
import {
  View,
  Text,
  Button,
} from 'react-native';
import { ModalPanel } from './modalPanel.component';
import {
  ModalComponentCloseProps,
  ModalService,
} from '../../../theme';

jest.useFakeTimers();

describe('@modal panel checks', () => {

  const showModalTestId: string = '@modal/show';
  const hideModalTestIdInner: string = '@modal/hide-inner';
  const hideModalTestIdOuter: string = '@modal/hide-outer';

  interface HooksProps {
    componentDidMount?: () => void;
    componentWillUnmount?: () => void;
  }

  class ModalPanelTest extends React.Component<HooksProps> {

    private modalId: string = '';

    public componentDidMount(): void {
      if (this.props.componentDidMount) {
        this.props.componentDidMount();
      }
    }

    public componentWillUnmount() {
      if (this.props.componentWillUnmount) {
        this.props.componentWillUnmount();
      }
    }

    public showModal() {
      this.modalId = ModalService.show(<TestModal onRequestClose={() => 1}/>, true);
    }

    public hideModal() {
      ModalService.hide(this.modalId);
    }

    public render() {
      return (
        <ModalPanel>
          <View>
            <Text>Modal Panel Test</Text>
          </View>
          <Button
            title='Open Modal'
            onPress={() => this.showModal()}
            testID={showModalTestId}
          />
          <Button
            title='Hide Modal'
            onPress={() => this.hideModal()}
            testID={hideModalTestIdOuter}
          />
        </ModalPanel>
      );
    }
  }

  class TestModal extends React.Component<ModalComponentCloseProps> {

    render(): React.ReactNode {
      return (
        <View>
          <Text>This is Test Modal</Text>
          <Button
            title='Close Modal'
            onPress={this.props.onCloseModal}
            testID={hideModalTestIdInner}
          />
        </View>
      );
    }
  }

  it('* modal panel renders properly', () => {
    const panel = render(<ModalPanelTest/>);
    expect(panel).toMatchSnapshot();
  });

  it('* modal panel render with props / children checking', () => {
    const panelChild = <View><Text>Test</Text></View>;
    const panel = <ModalPanelTest children={panelChild}/>;
    const renderedPanel = render(panel);

    expect(panel.props.children).toBe(panelChild);
    expect(renderedPanel).toMatchSnapshot();
  });

  it('* modal panel l/c-hooks checks', () => {
    const componentDidMount = jest.fn();
    const componentWillUnmount = jest.fn();

    const wrapper = render(
      <ModalPanelTest
        componentDidMount={componentDidMount}
        componentWillUnmount={componentWillUnmount}
      />,
    );
    expect(componentDidMount).toHaveBeenCalled();
    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalled();
  });

  it('* close modal checking inner', () => {
    const application = render(<ModalPanelTest/>);
    fireEvent.press(application.getByTestId(showModalTestId));
    fireEvent.press(application.getByTestId(hideModalTestIdInner));
    expect(application).toMatchSnapshot();
  });

  it('* close modal checking outer', () => {
    const application = render(<ModalPanelTest/>);
    fireEvent.press(application.getByTestId(showModalTestId));
    fireEvent.press(application.getByTestId(hideModalTestIdOuter));
    expect(application).toMatchSnapshot();
  });

});
