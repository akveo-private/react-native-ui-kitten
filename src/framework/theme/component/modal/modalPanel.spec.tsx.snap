// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@modal panel checks * close modal checking 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <Text>
      Modal Panel Test
    </Text>
  </View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="@modal/show"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Open Modal
      </Text>
    </View>
  </View>
</View>
`;

exports[`@modal panel checks * modal panel render with props / children checking 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <Text>
      Modal Panel Test
    </Text>
  </View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="@modal/show"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Open Modal
      </Text>
    </View>
  </View>
</View>
`;

exports[`@modal panel checks * modal panel renders properly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <Text>
      Modal Panel Test
    </Text>
  </View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="@modal/show"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Open Modal
      </Text>
    </View>
  </View>
</View>
`;
